Refactor networks data state to support place comparisons

rachelekm opened the pull request:
## Overview

In reviewing #693 I missed that we need all travel/route info for all user destinations to support comparisons in network details and compare tab. This PR restructure networks' time and paths data state shape so that routes and travel times are available for every destination as well as network mode at one time. This supports the [route and commute time comparisons in neighborhood detail.](https://github.com/azavea/echo-locator/issues/697#issuecomment-3228496318)

Along the way simplifies the networks data implementation so that neighborhood routes and travel times are derived as part of the thunk to fetch path and times data (vs individual selectors). This refactor is [following the discussion ](https://github.com/azavea/echo-locator/pull/693#issuecomment-3211535905) in #693 and allows for simplified state management and greater maintainability as we populate components with data.


### Checklist

- [x] `fixup!` commits have been squashed
- [x] `CHANGELOG.md` updated with summary of features or fixes, following
      [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) guidelines
- [x] `README.md` updated if necessary to reflect the changes
- [x] Run `./scripts/format` to lint, format, and fix the application source code.
- [x] CI passes after rebase

### Demo
New networks state structure:
- `networks` state no longer has parsed buffer responses from path and times endpoints
- Separate `timesAndRoutes` state in `networks` slice for derived route and travel time data, indexed by place label and then network mode:
<img width="1054" height="479" alt="Screenshot 2025-08-27 at 3 18 32 PM" src="https://github.com/user-attachments/assets/fe36fe5c-230a-4155-8a5d-11a3266dc2b8" />

- No regressions with ranked list for new network state setup
<img width="1034" height="281" alt="Screenshot 2025-08-27 at 10 30 05 AM" src="https://github.com/user-attachments/assets/8775477e-2f13-4a3e-bbe5-e8b935aa315a" />


## Testing Instructions

 * `scripts/server`
 * Open up discover page in browser and open dev tools to see console
 * Confirm `networks.networks` state no longer has array buffer values (`paths`, `targets`, and other keys like `stops` or `travelTimeSurface`)
 * Confirm new state in `networks` slice, `timesAndRoutesData`, populated with key that matches the active destination value in user profile
 * Confirm values of this destination key is route and travel time data organized for every network mode
 * Alongside this window open up production app and set values to defaults, set origin to 700 Boylston st, and mode to Peak
 * Compare ranked list in production to the ranked list in this branch's state and confirm there are no regressions


Resolves #697

rachelekm committed:
Update networks state shape for data by place

In addition to having times/paths data for each network mode to support commute time ranges, we need to have times/paths data of each mode for every destination to support place comparison view in designs.

To achieve this, separates paths and times data into separate networks state that can be organized by place.

rachelekm committed:
Refactor source code implementation for stability

Allows multiple time travel numbers and routes to be accessible in state for each network mode. Prevents large parsed time/path array buffers from being stored in state as well as adds derived route and travel time information to state, both of which will allow for greater code maintainability moving forward. This implementation prevents re-fetching path and times data on destination or network mode change.

rachelekm committed:
Update activeDest to more unique identifier

rachelekm committed:
Update CHANGELOG

aaronxsu commented on src/frontend/src/reducers/networks/utils/createNetworkNeighborhoodRoutes.ts:
```diff
@@ -0,0 +1,74 @@
+    NeighborhoodRoutePaths,
+    NeighborhoodRoutes,
+    NetworkAndTimeAndPathsData,
+} from "../types";
```
I'd move these imports into the import line above, since all of them import from the same module.

rachelekm replied:
Updated in eb029d844da8a7f8ca827631df8efaa77f7d46db

aaronxsu commented on src/frontend/src/reducers/networks/utils/createNetworkNeighborhoodRoutes.ts:
```diff
@@ -0,0 +1,74 @@
+    neighborhoods: Neighborhoods
+): NeighborhoodRoutes {
+    const routes: NeighborhoodRoutes = [];
+    neighborhoods.features.map((neighborhood, neighborhoodIndex) => {
```
This might be from the old codebase or Taui, but the use of `map` here seems a bit odd to me, since it uses the side-effect here to create `routes`- I'd either replace it by a `reduce` to have it ready in one go or a `forEach` as a proper loop.

rachelekm replied:
Updated in eb029d844da8a7f8ca827631df8efaa77f7d46db

aaronxsu commented on src/frontend/src/reducers/networks/utils/createNetworkNeighborhoodTravelTimes.ts:
```diff
@@ -0,0 +1,22 @@
+            network,
+            lonlat(neighborhood.geometry.coordinates)
+        );
+        return surface?.data ? surface.data[idx] : 0;
```
Seems like if `surface?.data` is false-y, then we could return 0 early and skip the calculation for index. This might have performance implications. I'd suggest we improve it. For example:

```diff
diff --git a/src/frontend/src/reducers/networks/utils/createNetworkNeighborhoodTravelTimes.ts b/src/frontend/src/reducers/networks/utils/createNetworkNeighborhoodTravelTimes.ts
index 1374ba3..ea6db02 100644
--- a/src/frontend/src/reducers/networks/utils/createNetworkNeighborhoodTravelTimes.ts
+++ b/src/frontend/src/reducers/networks/utils/createNetworkNeighborhoodTravelTimes.ts
@@ -13,10 +13,11 @@ export default function createNeighborhoodTravelTimes(
 ): number[] {
     return neighborhoods.features.map(neighborhood => {
         const surface = network.travelTimeSurface;
+        if (!surface?.data) return 0;
         const idx = coordinateToIndex(
             network,
             lonlat(neighborhood.geometry.coordinates)
         );
-        return surface?.data ? surface.data[idx] : 0;
+        return surface.data[idx];
     });
 }
```

rachelekm replied:
Updated in eb029d844da8a7f8ca827631df8efaa77f7d46db

aaronxsu commented on src/frontend/src/reducers/networks/networksThunk.ts:
```diff
@@ -54,65 +57,105 @@ export const getNetworks = createAsyncThunk(
-        } catch (error: any) {
-            return rejectWithValue(error.message);
+                        const travelTimeSurface = parseTimesData(
+                            timesResponse.value
```
Looks like there was a check for `pathsResponse` before parsing above- should we do something similar here?

rachelekm replied:
That check is to handle the empty Promise in the event the network mode is car (since there is no paths data for this mode). But we do expect a times data response for every network mode or else it should land in the try-catch to error.

aaronxsu commented on src/frontend/src/reducers/networks/networksThunk.ts:
```diff
@@ -54,65 +57,105 @@ export const getNetworks = createAsyncThunk(
+                            ...networkDetails,
+                            name: network as NetworkModeOptionKey,
+                            ...pathsData,
+                            travelTimeSurface: travelTimeSurface,
```
This could be simply `travelTimeSurface,`

rachelekm replied:
Updated in eb029d844da8a7f8ca827631df8efaa77f7d46db

aaronxsu commented on src/frontend/src/reducers/networks/networksThunk.ts:
```diff
@@ -54,65 +57,105 @@ export const getNetworks = createAsyncThunk(
+                    })
+                );
+                return {
+                    label: label,
```
After we have the user sign-in and profile hooked up with the backend API, I'd suggest we use the destination record's ID instead of the label (currently name or address here), because IDs are usually unique identifiers. If we are going with this solution, I'd suggest we add a TODO here.

rachelekm replied:
Added this comment in c670861f359d18f80e79031c982e2f2391baedc4

aaronxsu reviewed (approved):
Works as advertised. I left some minor suggestions for potential improvements.

rachelekm commented:
Thanks for the review and these suggestions, @aaronxsu! I implemented all of the above and rebased with develop. Will go ahead and merge on successful CI.
